### PR TITLE
Fix/resolve filenames

### LINKED ISSUE #81: context.textFile does not support hdfs
Hello,

I am using both S3 and HDFS and found a discrepancy between their `resolve_filenames` method in pysparkling because of the underlying libraries:

`boto.s3.bucket.Bucket.list` expect a path without a leading `/`. It is correctly called in `pysparkling.fileio.s3` by `S3.resolve_filenames(path="s3://bucket-name/data/part-00000.gz")` which triggers `bucket.list("data/part-00000.gz")` and works as expected according to RFC 8089 on [file URI scheme](https://en.wikipedia.org/wiki/File_URI_scheme).

When using hdfs a similar call is made to `hdfs.client.Client.list` but this function requires a leading `/` to have the same behaviour,

Currently `Hdfs.resolve_filenames("hdfs://hdfs-cluster.com/data/part-00000.gz")` calls `client.list("data/part-00000.gz")` which looks for the content of the folder `/user/username/data/'.

I'm creating a PR to align `Hdfs.resolve_filenames` on `S3.resolve_filenames`, it fixes `context.textFile` which does not work with hdfs as it uses Hdfs.resolve_filenames which return consider all paths as relative to the user home directory but then open them using `TextFile(file_path)` which expect an absolute path.

Moreover, due to the implementation of `Hdfs.resolve_filenames` it does not support patterns as the first wildcard is removed from the requested path expression by the tokenizer when checking if it matches the actual file path.

### PR DESCRIPTION
**Pull request by tools4origins**
This PR fixes #81

### CODE REVIEW

File: pysparkling/fileio/fs/hdfs.py [hunk 1]
```diff
@@ -65,27 +65,18 @@ def resolve_filenames(expr):
...
-        if not file_expr:
-            if '/' in fixed_path:
-                file_expr = fixed_path[fixed_path.rfind('/') + 1:]
-                fixed_path = fixed_path[:fixed_path.rfind('/')]
-            else:
-                file_expr = fixed_path
-                fixed_path = ''
+        fixed_path = "/" + t.next(['*', '?'])
+
+        if '/' in fixed_path:
```
Comment (svenkreiss) [inline pysparkling/fileio/fs/hdfs.py]: Isn't that always true from line above?

File: pysparkling/tests/test_resolve_filenames.py [hunk 2]
```diff
@@ -33,7 +56,36 @@ def test_s3_1():
             'crawl-data/CC-MAIN-2015-11/warc.paths.gz' in filenames)
 
 
+def test_hdfs_resolve_filenames():
+    from pysparkling.fileio.fs import Hdfs
+    unused_path = "/"
+    path = "hdfs://hdfs-cluster.com/user/username/input/part-*.gz"
```
Comment (svenkreiss) [inline pysparkling/tests/test_resolve_filenames.py]: Can you add a test for a path without a `*` or `?`?
And another test for omitting the `/part-*.gz`?

Comment (tools4origins) [reply pysparkling/tests/test_resolve_filenames.py]: Adding tests without `/part-*.gz` allowed me to notice a misalignment with pyspark which support both `hdfs://hdfs-cluster.com/user/username/input/` and `hdfs://hdfs-cluster.com/user/username/input`,

I modified the code and tests but it increase a bit the complexity

It still does not support `hdfs://hdfs-cluster.com/user/username/*/part-*.gz` but I believe that that should be another PR

File: pysparkling/utils.py [hunk 3]
```diff
@@ -24,3 +28,18 @@ def next(self, separator=None):
         value = self.expression[:sep_pos]
         self.expression = self.expression[sep_pos + len(separator):]
```
Comment (tools4origins) [inline pysparkling/utils.py]: This PR also fixes a bug in Tokenizer.next() is called with a list: `len(separator)` is the size of the list and not the length of the found separator. Hence `self.expression` is altered in a incorrect way, leading to an unexpected behaviour when called once again,

This use case (call with a list and then another call) only appear in fs.hdfs so there is no side effect in other modules

**Pull request merged**